libraries/amqp/reliable_amqp_publisher.cpp

```diff
@@ -0,0 +1,272 @@
+
+struct reliable_amqp_publisher_handler;
+
+struct reliable_amqp_publisher_impl {
+   reliable_amqp_publisher_impl(const std::string& url, const std::string& exchange, const std::string& routing_key,
+                                const boost::filesystem::path& unconfirmed_path, const std::optional<std::string>& message_id);
+   ~reliable_amqp_publisher_impl();
+   void pump_queue();
+   void publish_message_raw(std::vector<char>&& data);
+   void publish_messages_raw(std::deque<std::pair<std::string, std::vector<char>>>&& queue);
+   void publish_message_direct(const std::string& routing_key, std::vector<char> data);
+   bool verify_max_queue_size();
+
+   void channel_ready(AMQP::Channel* channel);
+   void channel_failed();
+
+   AMQP::Channel* channel = nullptr;  //nullptr when channel is not up
+   std::atomic_bool stopping = false;
+
+   unsigned in_flight = 0;
+   fc::unsigned_int batch_num = 0;
+   struct amqp_message {
+      fc::unsigned_int  num = 0; ///< unique numbers indicates amqp transaction set (reset on clean restart)
+      std::string       routing_key;
+      std::vector<char> data;
+   };
+   std::deque<amqp_message> message_deque;
+
+   std::thread thread;
+   boost::asio::io_context ctx;
+
+   single_channel_retrying_amqp_connection retrying_connection;
+
+   const boost::filesystem::path data_file_path;
+
+   const std::string exchange;
+   const std::string routing_key;
+   const std::optional<std::string> message_id;
+   bool logged_exceeded_max_depth = false;
+
+   boost::asio::strand<boost::asio::io_context::executor_type> user_submitted_work_strand = boost::asio::make_strand(ctx);
+};
+
+reliable_amqp_publisher_impl::reliable_amqp_publisher_impl(const std::string& url, const std::string& exchange, const std::string& routing_key,
+                                                           const boost::filesystem::path& unconfirmed_path, const std::optional<std::string>& message_id) :
+  retrying_connection(ctx, url, [this](AMQP::Channel* c){channel_ready(c);}, [this](){channel_failed();}),
+  data_file_path(unconfirmed_path), exchange(exchange), routing_key(routing_key), message_id(message_id) {
+
+   boost::system::error_code ec;
+   boost::filesystem::create_directories(data_file_path.parent_path(), ec);
+
+   if(boost::filesystem::exists(data_file_path)) {
+      try {
+         fc::datastream<fc::cfile> file;
+         file.set_file_path(data_file_path);
+         file.open("rb");
+         fc::raw::unpack(file, message_deque);
+         if( !message_deque.empty() )
+            batch_num = message_deque.back().num;
+      } FC_RETHROW_EXCEPTIONS(error, "Failed to load previously unconfirmed AMQP messages from ${f}", ("f", (fc::path)data_file_path));
+   }
+   else {
+      boost::filesystem::ofstream o(data_file_path);
+      FC_ASSERT(o.good(), "Failed to create unconfirmed AMQP message file at ${f}", ("f", (fc::path)data_file_path));
+   }
+   boost::filesystem::remove(data_file_path, ec);
```

Comment:
If preserving unsent messages is important, Instead of removing file here can  it be renamed as a backup file or deleted later?

Reply:
I believe our current plan is to always persist to disk as we go along to support hard-failure cases like `kill -9`

Reply:
I just think of a case that the file has some unconfirmed messages because of the previous problem.  And nodeos restarts and reads the file and crashed or killed by -9  before writing to a data file.  Since we deleted the data file already, the next rerun will not have a valid data file.

Reply:
(the reason it works this way) was IMO if nodeos crashes I feel like you're going to have bigger problems. You're going to have to start anew from a snapshot or reply/resync _anyways_. So you can pick a block to restart from that covers the area affected.

But as Kevin explains, there is a desire to harden usage of file further.